### PR TITLE
Feature: Improve qdevice/qnetd management

### PR DESCRIPTION
**Pull request by liangxin1300**
## Motivation
As the major HA configuration tool set, crmsh must support qdevice/qnetd management.

At currently master branch, crmsh already have some options to configure qdevice when doing ha-cluster-init process, but need to improve for validation, qnetd management,  making qdevice process as a bootstrap stage, non-TLS process and TLS certification process.

Note: `stage` here means run qdevice process separately, no need to re-run all bootstrap process

#### 1. Setup qdevice/qnetd during cluster init process
When deploying cluster service on first node `hanode1`, using
`crm cluster init --qnetd-hostname=qnetd-node`
(`qnetd-node` is the node used for qnetd service) can get a node running both cluster service and `corosync-qdevice` service
[related demo](https://asciinema.org/a/273458)
#### 2. Setup qdevice during cluster join process
Assuming first node `hanode1` finished init process with qdevice service running, on second node `hanode2`, running
`crm cluster join -c hanode1`
will get `corosync-qdevice` service running on `hanode2`
[related demo](https://asciinema.org/a/273461)
#### 3. Setup qdevice/qnetd on an running cluster
This also means configuring qdevice/qnetd could be a `stage` of bootstrap;
When cluster running on `hanode1` and `hanode2`, running
`crm cluster init qdevice --qnetd-hostname=qnetd-node`
Then `corosync-qdevice` service running on these two nodes.
[related demo](https://asciinema.org/a/273463)
#### 4. Remove qdevice from a cluster
running
`crm cluster remove --qdevice`
can remove qdevice configuration and stop corosync-qdevice service from all cluster nodes
[related demo](https://asciinema.org/a/273464)
#### 5. Show status
using
`crm corosync status [ring|quorum|qnetd]`
```
15sp1-1:~ # crm corosync status quorum 
Quorum information
------------------
Date:             Fri Oct 11 10:11:02 2019
Quorum provider:  corosync_votequorum
Nodes:            2
Node ID:          168430131
Ring ID:          168430131/12
Quorate:          Yes

Votequorum information
----------------------
Expected votes:   3
Highest expected: 3
Total votes:      3
Quorum:           2  
Flags:            Quorate Qdevice 

Membership information
----------------------
    Nodeid      Votes    Qdevice Name
 168430131          1    A,V,NMW 10.10.10.51 (local)
 168430132          1    A,V,NMW 10.10.10.52
         0          1            Qdevice
```
```
15sp1-1:~ # crm corosync status qnetd
Cluster "hacluster":
    Algorithm:		Fifty-Fifty split
    Tie-breaker:	Node with lowest node ID
    Node ID 168430131:
        Client address:		::ffff:10.10.10.51:37922
        Configured node list:	168430131, 168430132
        Membership node list:	168430131, 168430132
        Vote:			ACK (ACK)
    Node ID 168430132:
        Client address:		::ffff:10.10.10.52:48050
        Configured node list:	168430131, 168430132
        Membership node list:	168430131, 168430132
        Vote:			No change (ACK)
```
## Example in steps
1. Prepare three nodes(`node1`, `node2` and `node3`).
    `node1` and `node2` are for cluster, with `crmsh corosync corosync-qdevice` installed.
    `node3` is for qnetd, just with `corosync-qnetd` installed.

2. On `node1`, 
    run `ha-cluster-init --qnetd-hostname=node3` (or `crm cluster init --qnetd-hostname=node3`),
    after original init process and qdevice process finished,
    on `node2`, 
    run `ha-cluster-join` (or `crm cluster join`)

3. Then got a two nodes cluster with qdevice service running
#### For init process
```
    --qnetd-hostname=HOST        Host or IP of the QNetd server to be used
    --qdevice-port=PORT
                        TCP PORT of QNetd server(default:5403)
    --qdevice-algo=ALGORITHM
                        QNetd decision ALGORITHM(ffsplit/lms, default:ffsplit)
    --qdevice-tie-breaker=TIE_BREAKER
                        QNetd TIE_BREAKER(lowest/highest/valid_node_id,
                        default:lowest)
    --qdevice-tls=TLS   Whether using TLS on QDevice/QNetd(on/off/required,
                        default:on)
    --qdevice-heuristics=COMMAND
                        COMMAND to run with absolute path. For multiple
                        commands, use ";" to separate(details about heuristics
                        can see man 8 corosync-qdevice)
```
#### Unit test
using unittest + mock
#### Functional test
using python-behave + docker + travisCI

## Task check list
- [x] valid qdevice options
- [x] manage qnetd node
- [x] make qdevice process as a bootstrap stage
- [x] non-tls process
- [x] tls certification process
- [x] remove qdevice/qnetd
- [x] testing

### CODE REVIEW

**Comment by liangxin1300** [comment]
@diegoakechi @arbulu89 For this PR, major part of codes and unit tests have finished.
Please review if you have time.
After #483 merged, I will append more functional test cases for this PR based on `behave`

File: crmsh/bootstrap.py [hunk 1]
```diff
@@ -907,6 +911,15 @@ def valid_port(port, prev_value=None):
     return False
 
 
+def add_nodelist_from_cmaptool():
+    for nodeid, iplist in utils.get_nodeinfo_from_cmaptool().items():
+        try:
+            corosync.add_node_ucast(iplist, nodeid)
+        except ValueError as err:
+            if "was already configured" in str(err):
```
[the remark this answers is not in the document]
Comment (liangxin1300) [reply crmsh/bootstrap.py]: Thanks!
Changes in e31ba8995809995918059f8eb8667846e4c55bc3

File: crmsh/bootstrap.py [hunk 2]
```diff
@@ -1712,7 +1868,7 @@ def update_nodeid(nodeid, node=None):
         error("{} is not readable. Please ensure that hostnames are resolvable.".format(corosync.conf()))
 
     # if unicast, we need to add our node to $corosync.conf()
-    is_unicast = "nodelist" in open(corosync.conf()).read()
+    is_unicast = corosync.get_value("totem.transport") == "udpu"
```
[the remark this answers is not in the document]
Comment (liangxin1300) [reply crmsh/bootstrap.py]: Thanks!
Changes in a2a6c36a52805626d16723ddba031fc37981bfbc

File: crmsh/corosync.py [hunk 3]
```diff
@@ -58,17 +59,406 @@ def __repr__(self):
...
+    def qdevice_p12_on_local(self):
+        return "{}/nssdb/{}".format(self.qdevice_path, self.qdevice_p12_filename)
+
+    @property
+    def qdevice_p12_on_cluster(self):
+        return "{}/{}/{}".format(self.qdevice_path, self.cluster_node, self.qdevice_p12_filename)
+
+    def valid_attr(self):
+        if self.ip == utils.this_node() or \
+           self.ip in utils.ip_in_local():
```
[the remark this answers is not in the document]
Comment (liangxin1300) [reply crmsh/corosync.py]: Thanks!
Combine in one line, I think the length is acceptable:)
And I will learn about Pylint, maybe create a new PR to integrate it in crmsh

File: crmsh/bootstrap.py [hunk 4]
```diff
@@ -1646,6 +1724,84 @@ def join_ssh_merge(_cluster_node):
             if isinstance(result, parallax.Error):
                 warn("scp to {} failed ({}), known_hosts update may be incomplete".format(host, str(result)))
 
+def update_expected_votes():
```
[the remark this answers is not in the document]
Comment (liangxin1300) [reply crmsh/bootstrap.py]: In this PR, for this function, what I've done is just make update_expected_votes as outer function for reuse, I does not change anything inside the function.
I think this function has many impacts on the whole bootstrap process, it does work, maybe it's a better opportunity to refactor it in another new PR with `behave` running

**Comment by liangxin1300** [comment]
note:
stage remove
language syntax
help message

**Comment by liangxin1300** [comment]
@arbulu89 @diegoakechi please review this, thanks!

File: crmsh/bootstrap.py [hunk 5]
```diff
@@ -414,6 +414,10 @@ def check_prereqs(stage):
             warn("No watchdog device found. If SBD is used, the cluster will be unable to start without a watchdog.")
             warned = True
 
+    if stage == "qdevice":
+        if not _context.qdevice:
+            error("Miss qdevice related option(at least with --qdevice)")
```
Comment (arbulu89) [inline crmsh/bootstrap.py]: I would change this to:
`qdevice related options are missing (--qdevice option is mandatory, find for more information using --help)`

File: crmsh/utils.py [hunk 6]
```diff
@@ -2198,4 +2198,59 @@ def check_ssh_passwd_need(hosts):
...
+            return True
+        else:
+            return False
+
+
+def valid_port(port):
+    return int(port) >= 1024 and int(port) <= 65535
+
+
+def use_qdevice():
```
Comment (arbulu89) [inline crmsh/utils.py]: I don't like much this name `use_qdevice`. This function is to check if qdevice is configured, right?
I would rename to `is_qdevice_configured` or similar as it returns a bool actually.
But maybe I'm missing something

File: crmsh/bootstrap.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (arbulu89) [reply crmsh/bootstrap.py]: Ok, perfect! Let's follow up with this in a future PR

File: crmsh/bootstrap.py [hunk 7]
```diff
@@ -1646,6 +1722,84 @@ def join_ssh_merge(_cluster_node):
...
+    expected_votes = 0
+    while True:
+        rc, nodelist_text = utils.get_stdout("cibadmin -Ql --xpath '/cib/status/node_state'")
+        if rc == 0:
+            try:
+                nodelist_xml = etree.fromstring(nodelist_text)
+                nodelist = [n.get('uname') for n in nodelist_xml.xpath('//node_state') if n.get('remote_node') != 'true']
+                if len(nodelist) >= 2:
+                    break
+            except Exception:
```
Comment (arbulu89) [inline crmsh/bootstrap.py]: This general `Exception` is quite risky. I think we could have some more specific exception check here.
Anyway, so far is not critical.

File: crmsh/bootstrap.py [hunk 8]
```diff
@@ -1646,6 +1722,84 @@ def join_ssh_merge(_cluster_node):
...
+        sleep(1)
+
+    # Increase expected_votes
+    # TODO: wait to adjust expected_votes until after cluster join,
+    # so that we can ask the cluster for the current membership list
+    # Have to check if a qnetd device is configured and increase
+    # expected_votes in that case
+    use_qdevice = utils.use_qdevice()
+    if nodelist is None:
+        for v in corosync.get_values("quorum.expected_votes"):
```
Comment (arbulu89) [inline crmsh/bootstrap.py]: Let's try to have more meaningful name, Chnage `v` by `votes` maybe

File: crmsh/corosync.py [hunk 9]
```diff
@@ -58,17 +59,405 @@ def __repr__(self):
...
         self.ip = ip
         self.port = port
         self.algo = algo
         self.tie_breaker = tie_breaker
+        self.tls = tls
+        self.cluster_node = cluster_node
+        self.askpass = False
+        self.cmds = cmds
+
+        self.qnetd_service = "corosync-qnetd.service"
```
Comment (arbulu89) [inline crmsh/corosync.py]: I think we could put all of them as constants outside the `__init__`. They don't like that can be changed, right?

File: crmsh/utils.py [hunk 10]
```diff
@@ -2198,4 +2198,59 @@ def check_ssh_passwd_need(hosts):
...
+    from . import corosync
+    return corosync.get_value("quorum.device.model") == "net"
+
+
+def qdevice_tls_on():
```
Comment (arbulu89) [inline crmsh/utils.py]: The same, maybe change to `is_qdevice_tls_on`

File: crmsh/bootstrap.py (hunk 5 above)
Comment (liangxin1300) [reply crmsh/bootstrap.py]: Changed,
thanks!

File: crmsh/utils.py (hunk 6 above)
Comment (liangxin1300) [reply crmsh/utils.py]: changed!

File: crmsh/corosync.py (hunk 9 above)
Comment (liangxin1300) [reply crmsh/corosync.py]: OK
changed

File: crmsh/utils.py (hunk 10 above)
Comment (liangxin1300) [reply crmsh/utils.py]: changed

**Comment by liangxin1300** [comment]
> @liangxin1300 Great work!!
> I'm still missing some documentation in some new methods docstrings, and to make the large methods in smaller pieces.
> Besides that, I have added some comments about renaming and some other topics.
> 
> The changes are huge, so it's not possible to check everything, but you added a lot of tests, so I think it's enough

@arbulu89 Thanks for reviewing!
I will merge it if it looks have no further big defects:)

**Comment by liangxin1300** [comment]
Note: should add more debug messages during certification processes

**Comment by diegoakechi** [comment]
@liangxin1300 , @arbulu89  is out today, but I think we are good to go here. We can do minor fixes in a follow-up PR if needed.

**Comment by liangxin1300** [comment]
Append two more commits:

- 76a6f13: add log and debug messages on each certificate steps
- 000f2a7: replace `--qdevice` as `--qnetd-hostname`

**Pull request merged**